source/ocproute.go

```diff
@@ -270,14 +289,73 @@ func endpointsFromOcpRoute(ocpRoute *routeapi.Route, ignoreHostnameAnnotation bo
 	return endpoints
 }
 
-func targetsFromOcpRouteStatus(status routeapi.RouteStatus) endpoint.Targets {
+func (ors *ocpRouteSource) targetsFromOcpRouteStatus(status routeapi.RouteStatus) endpoint.Targets {
 	var targets endpoint.Targets
 
 	for _, ing := range status.Ingress {
-		if ing.RouterCanonicalHostname != "" {
+		if ing.RouterName != "" && ocpMajorVersion == ocpMajorVersion4 {
+			// Extract load balancer IPs/Hostnames for the route
+			target, err := ors.getRouterLoadBalancerTargets(ing.RouterName)
+			if err != nil {
+				log.Warn(err)
+			}
+			targets = append(targets, target...)
+		} else if ing.RouterCanonicalHostname != "" && ocpMajorVersion == ocpMajorVersion3 {
+			// Append RouterCanonicalHostname in case it's OCP v3
 			targets = append(targets, ing.RouterCanonicalHostname)
 		}
 	}
-
 	return targets
 }
+
+func (ors *ocpRouteSource) getRouterLoadBalancerTargets(routerName string) (endpoint.Targets, error) {
+	// LabelSelector for selecting services corresponding to router
+	labelSelector, err := metav1.ParseToLabelSelector(ingressControllerOwnerLabel + "=" + routerName)
+	if err != nil {
+		return nil, err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	listOptions := metav1.ListOptions{
+		LabelSelector: selector.String(),
+	}
+
+	// List of services filtered by owning ingress controller
+	services, err := ors.kubeClient.CoreV1().Services(ors.namespace).List(context.TODO(), listOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Iterate through services and find service of loadbalancer type
+	var targets endpoint.Targets
+	for _, service := range services.Items {
+		if service.Spec.Type == v1.ServiceTypeLoadBalancer {
+			// Add load balancer IPs/Hostnames against the route
+			targets = append(targets, ExtractLoadBalancerTargets(&service)...)
+		}
+	}
+	return targets, err
+}
+
+// Returns Openshift Container Platform Major Version
+func getOCPMajorVersion(kubeClient kubernetes.Interface) int {
+	// Since, IngressController was introduced in OCP v4.x we can verify if the current cluster is on OCP v4.x by
+	// ensuring that resource exists against Group(operator.openshift.io), Version(v1) and Kind(IngressController)
+	// In case it doesn't exist we assume that it's running on OCP v3.x
```

Comment:
Would it be probably better to define the OCPMajorVersion through the k8s version? OCP3.x is <=1.11, OCP4.x is >=1.12?

Reply:
Is that possible via the kube client ? Because as far as i have researched it isn't. The guideline is to not check for openshift/kubernetes versions but for what exactly you need i.e. specific version of API etc.